contracts/factory/MagicDropCloneFactory.sol

```diff
@@ -0,0 +1,245 @@
+    );
+
+    /*==============================================================
+    =                             ERRORS                           =
+    ==============================================================*/
+
+    error InitializationFailed();
+    error SaltAlreadyUsed();
+    error ContractAlreadyDeployed(address deployedAddress);
+    error RegistryAddressCannotBeZero();
+
+    /*==============================================================
+    =                          INITIALIZER                         =
+    ==============================================================*/
+
+    /// @notice Initializes the contract
+    /// @param initialOwner The address of the initial owner
+    /// @param registry The address of the registry contract
+    /// @dev This function can only be called once
+    function initialize(address initialOwner, address registry) public initializer {
+        if (registry == address(0)) {
+            revert RegistryAddressCannotBeZero();
+        }
+
+        _registry = MagicDropTokenImplRegistry(registry);
+        _initializeOwner(initialOwner);
+
+        emit MagicDropFactoryInitialized();
+    }
+
+    /*==============================================================
+    =                      PUBLIC WRITE METHODS                    =
+    ==============================================================*/
+
+    /// @notice Creates a new deterministic clone of a MagicDrop contract
+    /// @param name The name of the new contract
+    /// @param symbol The symbol of the new contract
+    /// @param standard The token standard of the new contract
+    /// @param initialOwner The initial owner of the new contract
+    /// @param implId The implementation ID
+    /// @param salt A unique salt for deterministic address generation
+    /// @return The address of the newly created contract
+    function createContractDeterministic(
+        string calldata name,
+        string calldata symbol,
+        TokenStandard standard,
+        address payable initialOwner,
+        uint32 implId,
+        bytes32 salt
+    ) external returns (address) {
+        address impl;
+        // Retrieve the implementation address from the registry
+        if (implId == 0) {
+            impl = _registry.getDefaultImplementation(standard);
+        } else {
+            impl = _registry.getImplementation(standard, implId);
+        }
+
+        /// @solidity memory-safe-assembly
+        assembly {
+            mstore(0x00, salt)
+            mstore(0x20, MAGICDROP_FACTORY_STORAGE)
+            let saltUsed := sload(keccak256(0x00, 0x40)) // usedSalts[salt]
+
+            if saltUsed {
+                // revert SaltAlreadyUsed()
+                mstore(0x00, 0x0ced3043)
+                revert(0x1c, 0x04)
+            }
+        }
+
+        // Predict the address where the contract will be deployed
+        address predictedAddress = LibClone.predictDeterministicAddress(impl, salt, address(this));
```

Comment:
It may be cheaper to do this off chain. There's a thread somewhere about gas efficiency of making comparisons vs just letting the tx fail. It would be interesting to see the gas usage of each.